src/playground-code-editor.ts

```diff
@@ -384,17 +411,252 @@ export class PlaygroundCodeEditor extends LitElement {
+          fileName: this.fileName,
+          fileContent: this.value,
+          tokenUnderCursor,
+          cursorIndex: this.cursorIndex,
+          provideCompletions: (completions: EditorCompletion[]) =>
+            this._onCompletionsProvided(id, completions),
+        },
+      })
+    );
+  }
+
+  private _onCompletionsProvided(id: number, completions: EditorCompletion[]) {
+    // To prevent race conditioning, check that the completions provided
+    // are from the latest completions request.
+    if (id !== this._currentCompletionRequestId) {
+      return;
+    }
+
+    this.completions = completions;
+  }
+
+  private _currentFiletypeSupportsCompletion() {
+    // Currently we are only supporting code completion for TS. Change
+    // this in a case that we start to support it for other languages too.
+    return this.type === 'ts';
+  }
+
   focus() {
     this._codemirrorEditable?.focus();
   }
 
+  private _hintFunction(
+    cm: Editor
+  ): Hints | null | undefined | PromiseLike<Hints | null | undefined> {
+    const cursorPosition = cm.getCursor('start');
+    const token = cm.getTokenAt(cursorPosition);
+    const lineNumber = cursorPosition.line;
+
+    const hintList = this._completionsAsHints();
+
+    const hints = {
+      from: {line: lineNumber, ch: token.start} as Position,
+      to: {line: lineNumber, ch: token.end} as Position,
+      list: hintList,
+    } as Hints;
+
+    CodeMirror.on(
+      hints,
+      'select',
+      async (hint: Hint | string, element: Element) => {
+        if (!this._isCodeEditorHint(hint)) return;
+        // If the current selection is the same, e.g. the completions were just
+        // updated by user input, instead of moving through compltions, we don't
+        // want to re-render and re-fetch the details.
+        if (this._currentSelectionLabel === hint.text) return;
+
+        this._onSelectedChange?.();
+
+        this._renderHint(element as HTMLElement, hints, hint, hint.details);
+      }
+    );
+
+    return hints;
+  }
+
+  private _isCodeEditorHint(hint: Hint | string): hint is CodeEditorHint {
+    return (
+      typeof hint !== 'string' &&
+      Object.prototype.hasOwnProperty.call(hint, 'details')
+    );
+  }
+
+  private _renderHint(
+    element: HTMLElement | undefined,
+    _data: Hints,
+    hint: CodeEditorHint,
+    detail?: Promise<EditorCompletionDetails>
+  ) {
+    if (!element) return;
+
+    const itemIndex = _data.list.indexOf(hint);
+    const completionData = this.completions?.[itemIndex];
+    const objectName = this._buildHintObjectName(
+      hint.displayText,
+      completionData
+    );
+    // Render the actual completion item first
+    this._renderCompletionItem(objectName, element);
+
+    // And if we have the detail promise passed into this function,
+    // we want to asynchronously update the detail info into our completion
+    // item. We don't want to block the rendering, so we don't use await.
+    if (detail !== undefined) {
```

Comment:
Would we ever not have a detail promise?

Reply:
The detail promise is passed into this function only for the item currently highlighted from the completion list.

I've added documentation regarding it above the check